### PR TITLE
Rebase revolution

### PR DESCRIPTION
**Pull request by Caleb-T-Owens**
I'm going to work on some documentation tomorrow.

For now, the key functions are:

```rs
    /// cherry-pick, but understands GitButler conflicted states
    ///
    /// cherry_pick_gitbutler should always be used in favour of libgit2 or gitoxide
    /// cherry pick functions
    fn cherry_pick_gitbutler(
        &self,
        head: &git2::Commit,
        to_rebase: &git2::Commit,
    ) -> Result<git2::Index, anyhow::Error> {
        // we need to do a manual 3-way patch merge
        // find the base, which is the parent of to_rebase
        let base = if to_rebase.is_conflicted() {
            self.find_real_tree(to_rebase, ConflictedTreeKey::Ours)?
        } else {
            let base_commit = to_rebase.parent(0)?;
            self.find_real_tree(&base_commit, Default::default())?
        };
        // Get the original ours
        let ours = self.find_real_tree(head, ConflictedTreeKey::Theirs)?;
        // Get the original theirs
        let thiers = self.find_real_tree(to_rebase, ConflictedTreeKey::Theirs)?;

        self.merge_trees(&base, &ours, &thiers, None)
            .context("failed to merge trees for cherry pick")
    }
```

Committing conflicted files as part of a rebase
```rs
fn commit_conflicted_cherry_result<'l>(
    ctx: &'l CommandContext,
    head: git2::Commit,
    to_rebase: git2::Commit,
    mut cherrypick_index: git2::Index,
) -> Result<git2::Commit<'l>> {
    let repository = ctx.repository();
    let commit_headers = to_rebase.gitbutler_headers();

    // If the commit we're rebasing is conflicted, use the commits original base.
    let base_tree = if to_rebase.is_conflicted() {
        repository.find_real_tree(&to_rebase, ConflictedTreeKey::Ours)?
    } else {
        let base_commit = to_rebase.parent(0)?;
        repository.find_real_tree(&base_commit, Default::default())?
    };

    // in case someone checks this out with vanilla Git, we should warn why it looks like this
    let readme_content =
        b"You have checked out a GitButler Conflicted commit. You probably didn't mean to do this.";
    let readme_blob = repository.blob(readme_content)?;

    // This is what can only be described as "a tad gross" but
    // AFAIK there is no good way of resolving conflicts in
    // an index without writing it *somewhere*
    let temporary_directory = tempdir().context("Failed to create temporary directory")?;
    let branch_name = Uuid::new_v4().to_string();
    let worktree = repository
        .worktree(
            &branch_name,
            &temporary_directory.path().join("repository"),
            None,
        )
        .context("Failed to create worktree")?;
    let worktree_repository =
        Repository::open_from_worktree(&worktree).context("Failed to open worktree repository")?;

    worktree_repository
        .set_index(&mut cherrypick_index)
        .context("Failed to set cherrypick index as worktree index")?;

    let mut conflicted_files = Vec::new();

    // get a list of conflicted files from the index
    let index_conflicts = cherrypick_index.conflicts()?.flatten().collect::<Vec<_>>();
    let mut theirs: Vec<git2::IndexEntry> = vec![];

    for conflict in index_conflicts {
        if let Some(their) = conflict.their {
            let path = std::str::from_utf8(&their.path).unwrap().to_string();
            conflicted_files.push(path);

            let data = repository.find_blob(their.id)?;
            let data = data.content();

            // For some reason we need to resolve the
            // conflicts using the "their" side and
            // then modify the tree afterwards.
            cherrypick_index
                .add_frombuffer(&their, data)
                .context("Failed to add resolution")?;

            theirs.push(their)
        }
    }

    let resolved_tree = cherrypick_index
        .write_tree_to(repository)
        .context("Failed to write cherry index")?;
    let resolved_tree = repository.find_tree(resolved_tree)?;
    let mut resolved_tree_updater = TreeUpdateBuilder::new();

    for their in theirs {
        resolved_tree_updater.upsert(their.path, their.id, git2::FileMode::Blob);
    }

    let resolved_tree_id = resolved_tree_updater.create_updated(repository, &resolved_tree)?;

    // convert files into a string and save as a blob
    let conflicted_files_string = conflicted_files.join("\n");
    let conflicted_files_blob = repository.blob(conflicted_files_string.as_bytes())?;

    // create a treewriter
    let mut tree_writer = repository.treebuilder(None)?;

    let side0 = repository.find_real_tree(&head, ConflictedTreeKey::Ours)?;
    let side1 = repository.find_real_tree(&to_rebase, ConflictedTreeKey::Theirs)?;

    // save the state of the conflict, so we can recreate it later
    tree_writer.insert(&*ConflictedTreeKey::Ours, side0.id(), 0o040000)?;
    tree_writer.insert(&*ConflictedTreeKey::Theirs, side1.id(), 0o040000)?;
    tree_writer.insert(&*ConflictedTreeKey::Base, base_tree.id(), 0o040000)?;
    tree_writer.insert(
        &*ConflictedTreeKey::AutoResolution,
        resolved_tree_id,
        0o040000,
    )?;
    tree_writer.insert(
        &*ConflictedTreeKey::ConflictFiles,
        conflicted_files_blob,
        0o100644,
    )?;
    tree_writer.insert("README.txt", readme_blob, 0o100644)?;

    let tree_oid = tree_writer.write().context("failed to write tree")?;

    let commit_headers = commit_headers.map(|commit_headers| {
        let conflicted_file_count = conflicted_files
            .len()
            .try_into()
            .expect("If you have more than 2^64 conflicting files, we've got bigger problems");
        CommitHeadersV2 {
            conflicted: Some(conflicted_file_count),
            ..commit_headers
        }
    });

    // write a commit
    let commit_oid = repository
        .commit_with_signature(
            None,
            &to_rebase.author(),
            &to_rebase.committer(),
            &to_rebase.message_bstr().to_str_lossy(),
            &repository
                .find_tree(tree_oid)
                .context("failed to find tree")?,
            &[&head],
            commit_headers,
        )
        .context("failed to create commit")?;

    // Tidy up worktree
    {
        temporary_directory.close()?;
        worktree.prune(None)?;
        repository
            .find_branch(&branch_name, git2::BranchType::Local)?
            .delete()?;
    }

    repository
        .find_commit(commit_oid)
        .context("failed to find commit")
}
```

Edit mode and rebasing have been moved to a project specific setting 
Screenshot 2024-08-22 at 18 09 14

### CODE REVIEW

**Review by krlvi — APPROVED** [review]
lgtm - lets try this and monitor

File: crates/gitbutler-repo/src/rebase.rs [hunk 1]
```diff
@@ -48,54 +52,224 @@ pub fn cherry_rebase_group(
...
+            commit_headers,
+        )
+        .context("failed to create commit")?;
+
+    repository
+        .find_commit(commit_oid)
+        .context("failed to find commit")
+}
+
+fn commit_conflicted_cherry_result<'l>(
```
Comment (krlvi) [inline crates/gitbutler-repo/src/rebase.rs]: btw why not the conventional life time names (ie. 'a 'b)?

Comment (Caleb-T-Owens) [reply crates/gitbutler-repo/src/rebase.rs]: :+1:

Comment (Caleb-T-Owens) [reply crates/gitbutler-repo/src/rebase.rs]: You're right. That was a very short and meaningless name.

```rs
fn commit_unconflicted_cherry_result<'repository>(
    ctx: &'repository CommandContext,
    head: git2::Commit<'repository>,
    to_rebase: git2::Commit,
    mut cherrypick_index: git2::Index,
) -> Result<git2::Commit<'repository>> {
```

File: crates/gitbutler-repo/src/rebase.rs [hunk 2]
```diff
@@ -128,12 +128,12 @@ fn commit_unconflicted_cherry_result<'repo>(
         .context("failed to find commit")
 }
 
-fn commit_conflicted_cherry_result<'l>(
-    ctx: &'l CommandContext,
+fn commit_conflicted_cherry_result<'repository>(
```
Comment (krlvi) [inline crates/gitbutler-repo/src/rebase.rs:131]: How do you feel about sticking to the rust conventions and idiomatic lifetime naming? It's technically not an abbreviation...so don't worry about calling it `'a`

Comment (Caleb-T-Owens) [reply crates/gitbutler-repo/src/rebase.rs:131]: Git2 has the lifetime named `'repo`

Screenshot 2024-08-22 at 18 21 12

Comment (Caleb-T-Owens) [reply crates/gitbutler-repo/src/rebase.rs:131]: The lifetime in this function is related to the repository, so it makes sense to name it accordingly

Comment (krlvi) [reply crates/gitbutler-repo/src/rebase.rs:131]: Fair enough 🤷. I am mostly trying to imitate the style of code in the standard lib.

**Pull request merged**